### PR TITLE
add some filename niceties

### LINKED ISSUE #43: `-` as alias for `/dev/stdin`
mipsy can read a file in two locations
1) command line
2) `load` command
both of these places have reasons to allow reading input from stdin. Mainly for debugging.
allowing `-` as an alias for `stdin` would be a nice time save.

### PR DESCRIPTION
**Pull request by joooooooooooooooooooooooooooooooooooosh**
fixes #43 and #151

### CODE REVIEW

**Comment by insou22** [comment]
we probably don't want to do tilde expansion in mipsy (non-interactive)
occasionally a user may want to run a file with a literal '~' in the name, which is why shells let you opt-out of expansion

**Comment by joooooooooooooooooooooooooooooooooooosh** [comment]
> we probably don't want to do tilde expansion in mipsy (non-interactive) occasionally a user may want to run a file with a literal '~' in the name, which is why shells let you opt-out of expansion
> 
> the rest of the changes (including ~ expansion in interactive) look good though!

They could technically still opt-out by doing `./file-name` but I see your point - if the filename passed in hasn't had the ~ expanded by the shell it's probably expected behaviour not to expand it ourselves.

File: crates/mipsy_utils/src/expand.rs [hunk 1]
```diff
@@ -7,10 +7,17 @@ pub fn expand_tilde<P: AsRef<Path> + ?Sized>(path: &'_ P) -> Cow<'_, Path> {
...
+    } else {
+        return path.into()
+    };
+
+    if path.starts_with("~/") {
         home.join(path.strip_prefix("~").unwrap()).into()
     } else {
-        path.into()
+        // TODO: support ~other_user/rest/of/path
+        // for now, assume this is a lone ~
```
Comment (insou22) [inline crates/mipsy_utils/src/expand.rs:20]: https://docs.rs/users might help

Comment (joooooooooooooooooooooooooooooooooooosh) [reply crates/mipsy_utils/src/expand.rs:20]: This doesn't seem to provide the ability to retrieve a user's home directory, only their uid, name, and groups.
The current Rust solution seems to be to interface with the C function `getpwnam_r`, not sure why the Users crate doesn't have this

Comment (joooooooooooooooooooooooooooooooooooosh) [reply crates/mipsy_utils/src/expand.rs:20]: wait nvm reading the source code it does provide this under OS_specific traits but this wasn't in the documentation

Comment (Dylan-Brotherston) [reply crates/mipsy_utils/src/expand.rs:20]: There is also
https://crates.io/crates/shellexpand
But it looks to be unmaintained.

**Comment by insou22** [comment]
awesome! just want to make sure it builds on windows before merging

**Comment by Dylan-Brotherston** [comment]
I doesn't
```
Compiling users v0.11.0
error[E0433]: failed to resolve: could not find `unix` in `os`
  --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:36:14
   |
36 | use std::os::unix::ffi::OsStrExt;
   |              ^^^^ could not find `unix` in `os`

error[E0432]: unresolved imports `libc::uid_t`, `libc::gid_t`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\lib.rs:128:16
    |
128 | pub use libc::{uid_t, gid_t};
    |                ^^^^^  ^^^^^ no `gid_t` in the root
    |                |
    |                no `uid_t` in the root

error[E0432]: unresolved imports `libc::uid_t`, `libc::gid_t`
  --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:45:20
   |
45 | use libc::{c_char, uid_t, gid_t, c_int};
   |                    ^^^^^  ^^^^^ no `gid_t` in the root
   |                    |
   |                    no `uid_t` in the root

error[E0432]: unresolved import `libc::passwd`
  --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:46:5
   |
46 | use libc::passwd as c_passwd;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ no `passwd` in the root

error[E0432]: unresolved import `libc::group`
  --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:47:5
   |
47 | use libc::group as c_group;
   |     ^^^^^^^^^^^^^^^^^^^^^^ no `group` in the root

error[E0432]: unresolved imports `libc::uid_t`, `libc::gid_t`
  --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\cache.rs:87:12
   |
87 | use libc::{uid_t, gid_t};
   |            ^^^^^  ^^^^^ no `gid_t` in the root
   |            |
   |            no `uid_t` in the root

error[E0432]: unresolved imports `libc::uid_t`, `libc::gid_t`
  --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\mock.rs:63:16
   |
63 | pub use libc::{uid_t, gid_t};
   |                ^^^^^  ^^^^^ no `gid_t` in the root
   |                |
   |                no `uid_t` in the root

error[E0432]: unresolved imports `libc::uid_t`, `libc::gid_t`
 --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\switch.rs:4:12
  |
4 | use libc::{uid_t, gid_t, c_int};
  |            ^^^^^  ^^^^^ no `gid_t` in the root
  |            |
  |            no `uid_t` in the root

error[E0432]: unresolved imports `libc::uid_t`, `libc::gid_t`
 --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\traits.rs:4:12
  |
4 | use libc::{uid_t, gid_t};
  |            ^^^^^  ^^^^^ no `gid_t` in the root
  |            |
  |            no `uid_t` in the root

error[E0433]: failed to resolve: could not find `UserExtras` in `os`
  --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:78:26
   |
78 |         let extras = os::UserExtras::default();
   |                          ^^^^^^^^^^ could not find `UserExtras` in `os`

error[E0433]: failed to resolve: could not find `GroupExtras` in `os`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:192:26
    |
192 |         let extras = os::GroupExtras::default();
    |                          ^^^^^^^^^^^ could not find `GroupExtras` in `os`

error[E0433]: failed to resolve: could not find `UserExtras` in `os`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:266:28
    |
266 |         extras:        os::UserExtras::from_passwd(passwd),
    |                            ^^^^^^^^^^ could not find `UserExtras` in `os`

error[E0433]: failed to resolve: could not find `GroupExtras` in `os`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:280:23
    |
280 |         extras:   os::GroupExtras::from_struct(group),
    |                       ^^^^^^^^^^^ could not find `GroupExtras` in `os`

error[E0412]: cannot find type `UserExtras` in module `os`
  --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:57:17
   |
57 |     extras: os::UserExtras,
   |                 ^^^^^^^^^^ not found in `os`

error[E0412]: cannot find type `GroupExtras` in module `os`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:171:17
    |
171 |     extras: os::GroupExtras,
    |                 ^^^^^^^^^^^ not found in `os`

error[E0425]: cannot find function `getpwuid_r` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:336:19
    |
336 |             libc::getpwuid_r(uid, &mut passwd, buf.as_mut_ptr(), buf.len(), &mut result)
    |                   ^^^^^^^^^^ not found in `libc`

error[E0425]: cannot find function `getpwnam_r` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:398:19
    |
398 |             libc::getpwnam_r(username.as_ptr(), &mut passwd, buf.as_mut_ptr(), buf.len(), &mut result)
    |                   ^^^^^^^^^^ not found in `libc`

error[E0425]: cannot find function `getgrgid_r` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:451:19
    |
451 |             libc::getgrgid_r(gid, &mut passwd, buf.as_mut_ptr(), buf.len(), &mut result)
    |                   ^^^^^^^^^^ not found in `libc`

error[E0425]: cannot find function `getgrnam_r` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:513:19
    |
513 |             libc::getgrnam_r(groupname.as_ptr(), &mut group, buf.as_mut_ptr(), buf.len(), &mut result)
    |                   ^^^^^^^^^^ not found in `libc`

error[E0425]: cannot find function `getuid` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:556:20
    |
556 |     unsafe { libc::getuid() }
    |                    ^^^^^^ help: a function with a similar name exists: `getpid`
    |
   ::: C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\libc-0.2.112\src\windows\mod.rs:474:5
    |
474 |     pub fn getpid() -> ::c_int;
    |     --------------------------- similarly named function `getpid` defined here

error[E0425]: cannot find function `geteuid` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:603:20
    |
603 |     unsafe { libc::geteuid() }
    |                    ^^^^^^^ help: a function with a similar name exists: `getpid`
    |
   ::: C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\libc-0.2.112\src\windows\mod.rs:474:5
    |
474 |     pub fn getpid() -> ::c_int;
    |     --------------------------- similarly named function `getpid` defined here

error[E0425]: cannot find function `getgid` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:647:20
    |
647 |     unsafe { libc::getgid() }
    |                    ^^^^^^ help: a function with a similar name exists: `getpid`
    |
   ::: C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\libc-0.2.112\src\windows\mod.rs:474:5
    |
474 |     pub fn getpid() -> ::c_int;
    |     --------------------------- similarly named function `getpid` defined here

error[E0425]: cannot find function `getegid` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:691:20
    |
691 |     unsafe { libc::getegid() }
    |                    ^^^^^^^ help: a function with a similar name exists: `getpid`
    |
   ::: C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\libc-0.2.112\src\windows\mod.rs:474:5
    |
474 |     pub fn getpid() -> ::c_int;
    |     --------------------------- similarly named function `getpid` defined here

error[E0425]: cannot find function `getgroups` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:745:15
    |
745 |         libc::getgroups(1024, buff.as_mut_ptr())
    |               ^^^^^^^^^ not found in `libc`

error[E0425]: cannot find value `buff` in this scope
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:783:21
    |
783 |     let mut count = buff.len() as c_int;
    |                     ^^^^ not found in this scope

error[E0425]: cannot find value `res` in this scope
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:799:8
    |
799 |     if res < 0 {
    |        ^^^ not found in this scope

error[E0425]: cannot find value `buff` in this scope
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:803:9
    |
803 |         buff.dedup();
    |         ^^^^ not found in this scope

error[E0425]: cannot find value `buff` in this scope
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:804:9
    |
804 |         buff.into_iter()
    |         ^^^^ not found in this scope

error[E0425]: cannot find function `setpwent` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:858:11
    |
858 |     libc::setpwent();
    |           ^^^^^^^^ not found in `libc`

error[E0425]: cannot find function `endpwent` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:873:24
    |
873 |         unsafe { libc::endpwent() };
    |                        ^^^^^^^^ not found in `libc`

error[E0425]: cannot find function `getpwent` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\base.rs:890:37
    |
890 |         let result = unsafe { libc::getpwent() };
    |                                     ^^^^^^^^ not found in `libc`

error[E0425]: cannot find function `setuid` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\switch.rs:41:26
    |
41  |     match unsafe { libc::setuid(uid) } {
    |                          ^^^^^^ help: a function with a similar name exists: `getpid`
    |
   ::: C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\libc-0.2.112\src\windows\mod.rs:474:5
    |
474 |     pub fn getpid() -> ::c_int;
    |     --------------------------- similarly named function `getpid` defined here

error[E0425]: cannot find function `setgid` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\switch.rs:72:26
    |
72  |     match unsafe { libc::setgid(gid) } {
    |                          ^^^^^^ help: a function with a similar name exists: `getpid`
    |
   ::: C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\libc-0.2.112\src\windows\mod.rs:474:5
    |
474 |     pub fn getpid() -> ::c_int;
    |     --------------------------- similarly named function `getpid` defined here

error[E0425]: cannot find function `seteuid` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\switch.rs:103:26
    |
103 |     match unsafe { libc::seteuid(uid) } {
    |                          ^^^^^^^ not found in `libc`

error[E0425]: cannot find function `setegid` in crate `libc`
   --> C:\Users\Dylan Brotherston\.cargo\registry\src\github.com-1ecc6299db9ec823\users-0.11.0\src\switch.rs:134:26
    |
134 |     match unsafe { libc::setegid(gid) } {
    |                          ^^^^^^^ not found in `libc`
```

**Comment by joooooooooooooooooooooooooooooooooooosh** [comment]
> awesome! just want to make sure it builds on windows before merging

1. do we want to expand the tilde on windows or just have it return the original string?
2. do we want to support a standard input equivalent for Windows when "-" is used or just leave it as a literal "-"?

**Comment by Dylan-Brotherston** [comment]
@joooooooooooooooooooooooooooooooooooosh looks like windows builds still aren't happy after c30ba5b

**Comment by insou22** [comment]
@joooooooooooooooooooooooooooooooooooosh 
https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies

File: crates/mipsy_utils/src/expand.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+use std::{borrow::Cow, path::Path};
+use dirs::home_dir;
+
+#[cfg(unix)]
+use users::os::unix::UserExt;
+#[cfg(bsd)]
+use users::os::bsd::UserExt;
+
+#[cfg(windows)]
```
Comment (insou22) [inline crates/mipsy_utils/src/expand.rs:9]: nit: this isn't windows specific behaviour, so `#[cfg(windows)]` is unnecessarily restrictive.
this should probably be `#[cfg(not(any(unix, bsd)))]`
iirc the only other target family is `wasm`, so my guess is mipsy_web wouldn't be able to build with this change

**Review by insou22 — CHANGES_REQUESTED** [review]

File: crates/mipsy_utils/src/expand.rs (hunk 2 above)
Comment (joooooooooooooooooooooooooooooooooooosh) [reply crates/mipsy_utils/src/expand.rs:9]: hmm looks like `bsd` isn't even a target family, I'll just do `cfg(unix)` and `cfg(not(unix))`

Comment (joooooooooooooooooooooooooooooooooooosh) [reply crates/mipsy_utils/src/expand.rs]: mipsy_web looks to be building and operating fine

**Review by insou22 — APPROVED** [review]

**Pull request merged**